permission/permission.go

```diff
@@ -0,0 +1,863 @@
+package permission
+
+import (
+	"crypto/ecdsa"
+	"encoding/json"
+	"errors"
+	"github.com/ethereum/go-ethereum/cmd/utils"
+	"github.com/ethereum/go-ethereum/raft"
+	"github.com/ethereum/go-ethereum/rpc"
+	"io/ioutil"
+	"math/big"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/eth"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethereum/go-ethereum/params"
+	pbind "github.com/ethereum/go-ethereum/permission/bind"
+)
+
+type NodeOperation uint8
+
+const (
+	NodeAdd NodeOperation = iota
+	NodeDelete
+)
+
+// permission config for bootstrapping
+type PermissionLocalConfig struct {
+	UpgrdAddress   string `json:"upgrdableAddress"`
+	InterfAddress  string `json:"interfaceAddress"`
+	ImplAddress    string `json:"implAddress"`
+	NodeAddress    string `json:"nodeMgrAddress"`
+	AccountAddress string `json:"accountMgrAddress"`
+	RoleAddress    string `json:"roleMgrAddress"`
+	VoterAddress   string `json:"voterMgrAddress"`
+	OrgAddress     string `json:"orgMgrAddress"`
+	NwAdminOrg     string `json:"nwAdminOrg"`
+	NwAdminRole    string `json:"nwAdminRole"`
+	OrgAdminRole   string `json:"orgAdminRole"`
+
+	Accounts      []string `json:"accounts"` //initial list of account that need full access
+	SubOrgBreadth string   `json:"subOrgBreadth"`
+	SubOrgDepth   string   `json:"subOrgDepth"`
+}
+
+type PermissionCtrl struct {
+	node       *node.Node
+	ethClnt    *ethclient.Client
+	eth        *eth.Ethereum
+	key        *ecdsa.PrivateKey
+	dataDir    string
+	permUpgr   *pbind.PermUpgr
+	permInterf *pbind.PermInterface
+	permNode   *pbind.NodeManager
+	permAcct   *pbind.AcctManager
+	permRole   *pbind.RoleManager
+	permOrg    *pbind.OrgManager
+	permConfig *types.PermissionConfig
+	orgChan    chan struct{}
+	nodeChan   chan struct{}
+	roleChan   chan struct{}
+	acctChan   chan struct{}
+	mux        sync.Mutex
+}
+
+// converts local permissions data to global permissions config
+func populateConfig(config PermissionLocalConfig) types.PermissionConfig {
```

Comment:
Please consider to implement `UnmarshalJSON()` to customize the transformation

Reply:
have a look at:

https://github.com/jpmorganchase/quorum/blob/1f0da48be656997cd8b446d40bee8702df92af27/common/types.go#L256-L259

https://github.com/jpmorganchase/quorum/blob/1f0da48be656997cd8b446d40bee8702df92af27/common/types_test.go#L93